earth2studio/statistics/rank.py

```diff
@@ -0,0 +1,157 @@
+            )
+
+        self.ensemble_dimension = ensemble_dimension
+        self._reduction_dimensions = reduction_dimensions
+        self.number_of_bins = number_of_bins
+
+    def __str__(self) -> str:
+        return "rank_histogram"
+
+    @property
+    def reduction_dimensions(self) -> list[str]:
+        return [self.ensemble_dimension] + self._reduction_dimensions
+
+    def __call__(
+        self,
+        x: torch.Tensor,
+        x_coords: CoordSystem,
+        y: torch.Tensor,
+        y_coords: CoordSystem,
+    ) -> tuple[torch.Tensor, CoordSystem]:
+        """
+        Apply metric to data `x` and `y`, checking that their coordinates
+        are broadcastable. While reducing over `reduction_dims`.
+
+        Parameters
+        ----------
+        x : torch.Tensor
+            Input tensor #1 intended to apply metric to.
+        x_coords : CoordSystem
+            Ordered dict representing coordinate system that describes the `x` tensor.
+            'reduction_dims' must be in coords.
+        y : torch.Tensor
+            Input tensor #2 intended to apply statistic to.
+        y_coords : CoordSystem
+            Ordered dict representing coordinate system that describes the `y` tensor.
+            'reduction_dims' must be in coords.
+
+        Returns
+        -------
+        tuple[torch.Tensor, CoordSystem]
+            Returns rank histogram tensor with appropriate reduced coordinates.
+        """
+        if not all([rd in x_coords for rd in self.reduction_dimensions]):
+            raise ValueError(
+                "Initialized reduction dimension does not appear in passed coords"
+            )
+
+        # Do some coordinate checking
+        # Assume ensemble_dim is in x_coords but not y_coords
+        if self.ensemble_dimension in y_coords:
+            raise ValueError(
+                f"{self.ensemble_dimension} should not be in y_coords but is."
+            )
+        if x.ndim != y.ndim + 1:
+            raise ValueError(
+                "x and y must have broadcastable shapes but got"
+                + f"{x.shape} and {y.shape}"
+            )
+        # Input coordinate checking
+        coord_count = 0
+        for c in x_coords:
+            if c != self.ensemble_dimension:
+                handshake_dim(y_coords, c, coord_count)
+                handshake_coords(y_coords, x_coords, c)
+                coord_count += 1
+
+        dim = list(x_coords).index(self.ensemble_dimension)
+
+        _ranks = torch.mean(1.0 * torch.ge(y, x), dim=dim)
+
+        # Reshape ranks using reduction dimensions
+        dims = [list(y_coords).index(rd) for rd in self._reduction_dimensions]
+        new_dims = list(range(len(dims)))
+        _ranks = torch.flatten(
+            torch.movedim(_ranks, dims, new_dims),
+            start_dim=new_dims[0],
+            end_dim=new_dims[-1],
+        )
+
+        # Compute histogram
+        _bin_edges = linspace(
```

Comment:
Similarly, bin edges should be the sorted values of the ensemble data. Does the above operation somehow transform this to bins between [0,1]?

Reply:
Yes, the ranking process transforms the data into 0 - 1.